# 线性判别分析

## 实验内容
1. 使用线性判别分析完成Breast_Cancer_Wisconsin结果预测问题。
2. 计算十折交叉验证下的精度(accuracy)，查准率(precision)，查全率(recall)，F1值。

## 评测指标  
1. 精度
2. 查准率
3. 查全率
4. F1

# 1. 读取数据

In [1]:
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv('breast-cancer.csv')


# 2. 导入模型

In [3]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import f1_score
from sklearn.model_selection import cross_val_predict
from sklearn import preprocessing

# 3. 提取数据

In [4]:
data = data.values 
data_x = data[:,2:-1]
data_y = data[:,1:2]

In [5]:
data_y = np.reshape(data_y,(-1))
data_y.shape

(569,)

# 4. 训练

请使用预处理后的数据进行模型的训练与预测

In [6]:
# 创建线性判别分析模型
model = LinearDiscriminantAnalysis()

# 十折交叉验证预测
prediction = cross_val_predict(model, data_x, data_y, cv = 10)


# 5. 评价指标的计算

计算模型的四项指标

In [7]:
accuracy = accuracy_score(data_y, prediction)
precision = precision_score(data_y, prediction, pos_label = 'M')
recall = recall_score(data_y, prediction, pos_label = 'M')
f1 = f1_score(data_y, prediction, pos_label = 'M')

print('accuracy:', accuracy)
print('precision:', precision)
print('recall:', recall)
print('f1:', f1)


accuracy: 0.9578207381370826
precision: 0.9895833333333334
recall: 0.8962264150943396
f1: 0.9405940594059405


###### 双击此处填写

数据集|精度|查准率|查全率|F1
-|-|-|-|-
Breast_Cancer_Wisconsin | 0.9578 | 0.9896 | 0.8962 | 0.9406


# 6. 选做：尝试对特征进行变换、筛选、组合后，训练模型并计算十折交叉验证后的四项指标

In [8]:
def evaluate_model(x, model):
    prediction = cross_val_predict(model, x, data_y, cv = 10)
    acc = accuracy_score(data_y, prediction)
    prec = precision_score(data_y, prediction, pos_label = 'M')
    rec = recall_score(data_y, prediction, pos_label = 'M')
    f1 = f1_score(data_y, prediction, pos_label = 'M')
    print('accuracy:', acc)
    print('precision:', prec)
    print('recall:', rec)
    print('f1:', f1)
    print()

# 模型1：原始全特征
model1 = LinearDiscriminantAnalysis()
print('模型1：原始全特征')
evaluate_model(data_x, model1)

# 模型2：标准化全特征
model2 = LinearDiscriminantAnalysis()
x_scaled = preprocessing.scale(data_x)
print('模型2：标准化全特征')
evaluate_model(x_scaled, model2)

# 模型3：仅使用前10个均值类特征
model3 = LinearDiscriminantAnalysis()
x_mean = data_x[:, :10]
print('模型3：仅使用前10个均值类特征')
evaluate_model(x_mean, model3)


模型1：原始全特征
accuracy: 0.9578207381370826
precision: 0.9895833333333334
recall: 0.8962264150943396
f1: 0.9405940594059405

模型2：标准化全特征
accuracy: 0.9578207381370826
precision: 0.9895833333333334
recall: 0.8962264150943396
f1: 0.9405940594059405

模型3：仅使用前10个均值类特征
accuracy: 0.9332161687170475
precision: 0.953125
recall: 0.8632075471698113
f1: 0.9059405940594059



###### 双击此处填写
1. 模型1的处理流程：原始全特征（29 个特征）
2. 模型2的处理流程：全特征做 z-score 标准化
3. 模型3的处理流程：仅使用前 10 个均值类特征

模型|数据集|精度|查准率|查全率|F1
-|-|-|-|-|-
模型1 | Breast_Cancer_Wisconsin | 0.9578 | 0.9896 | 0.8962 | 0.9406
模型2 | Breast_Cancer_Wisconsin | 0.9578 | 0.9896 | 0.8962 | 0.9406
模型3 | Breast_Cancer_Wisconsin | 0.9332 | 0.9531 | 0.8632 | 0.9059
